### ISSUE TITLE
[feature request] decouple creator plugins from API

### ISSUE THREAD

**Issue by pixelmord** [tinacms/tinacms#1147]
## Summary

See the discussion on the forum: https://community.tinacms.org/t/how-to-create-new-file-in-next-js-with-github-api/150

tl;dr;
ATM there's a RemarkCreatorPlugin in the gatsby package that works with the GIT-API and there's a MarkdownCreatorPlugin in the tinacms.org repo that works with the Github-API. It would be great to decouple those very similar plugins from the API they submit to.
So one way to accomplish that would be to pass in the onSubmit handler when instantiating the plugin as @kendallstrautman suggested.
Or even think of a way that APIs, in general, should implement an interface with a `submitFile` method that can be called by the creator plugin independent of which API is used.
The latter would make it a bit easier for users to not deal with the intricacies of either markdown or JSON file saving and they might only have to specify the desired API to be used in the creator plugin instance...

### Basic example

If the proposal involves a new or changed API, include a basic code example. Omit this section if it's not applicable.

**Example with passed submitHandler (pseudo code)**

```javascript
export const BlogPostCreatorPlugin = new MarkdownCreatorPlugin({
  label: 'New Blog Post',
  filename: form => {
    const slug = slugify(form.title)
    return `content/blog/${slug}.md`
  },
  fields: [
    // fields here
  ],
  frontmatter: postInfo => ({
    title: postInfo.title,
    date: moment(postInfo.date ? postInfo.date : new Date()).format(),
    author: postInfo.author ? postInfo.author : `Jane Doe`,
  }),
  body: () => `New post, who dis?`,
  afterCreate: response => {
    let url = fileToUrl(response.content.path.split('content')[1], 'blog')
    window.location.href = `/blog/${url}`
  },
  submitHandler: (form: FormShape, cms: CMS) => {
    // prepare form data
    const fileRelativePath = await this.filename(form);
    // ...
	// use api to save/create
    cms.api.MYAPI.doesItsThing();
  }
});

export class MarkdownCreatorPlugin<FormShape = any, FrontmatterShape = any>
  implements AddContentPlugin<FormShape> {
  __type: 'content-creator' = 'content-creator'
  name: AddContentPlugin<FormShape>['name']
  fields: AddContentPlugin<FormShape>['fields']

  afterCreate: (response: any) => void

  // Markdown Specific
  filename: (form: FormShape) => MaybePromise<string>
  frontmatter: (form: FormShape) => MaybePromise<FrontmatterShape>
  body: (form: any) => MaybePromise<string>

  submitHandler: (form: FormShape, cms: CMS) => MaybePromise<any>

  constructor(
    options: CreateMarkdownButtonOptions<FormShape, FrontmatterShape>
  ) {
    if (!options.filename) {
      console.error(MISSING_FILENAME_MESSAGE)
      throw new Error(MISSING_FILENAME_MESSAGE)
    }
    if (!options.submitHandler) {
      console.error(MISSING_SUBMITHANDLER_MESSAGE)
      throw new Error(MISSING_SUBMITHANDLER_MESSAGE)
    }

    if (!options.fields || options.fields.length === 0) {
      console.error(MISSING_FIELDS_MESSAGE)
      throw new Error(MISSING_FIELDS_MESSAGE)
    }

    this.name = options.label
    this.fields = options.fields
    this.filename = options.filename
    this.submitHandler = options.submitHandler
    this.frontmatter = options.frontmatter || (() => ({} as FrontmatterShape))
    this.body = options.body || (() => '')
    this.afterCreate = options.afterCreate || null
  }

  async onSubmit(form: FormShape, cms: CMS) {
    return await this.submitHandler(form, cms);
  }
}


```

### Motivation

Expected outcome would be a more universal use of the plugins and a more versatile application API
